src/js/core/UpdateStream.ts

```diff
@@ -0,0 +1,145 @@
+import { Observable } from "rxjs/Observable";
+import compareVersions from "compare-versions";
+import { request, RequestResponse } from "@dcos/http-service";
+import { BehaviorSubject } from "rxjs";
+
+import UserSettingsStore from "#SRC/js/stores/UserSettingsStore";
+import Util from "#SRC/js/utils/Util";
+import { SAVED_STATE_KEY } from "#SRC/js/constants/UserSettings";
+
+export const UpdateStreamType: symbol = Symbol("UpdateStreamType");
+const CHECK_DELAY: number = 24 * 60 * 60 * 1000; // Once every 24 hours
+const DCOS_UI_VERSION_SETTING: string = "dcosUIVersion";
+const DISMISSED_VERSION: string = "dismissedVersion";
+const LAST_TIME_CHECK: string = "lastTimeCheck";
+
+interface ContentTypeObject {
+  action: string;
+  actionType: string;
+  entity: string;
+  version: string;
+}
+
+interface DCOSUserSettings {
+  dcosUIVersion: DCOSUIVersion;
+  [key: string]: DCOSUIVersion;
+}
+
+interface DCOSUIVersion {
+  dismissedVersion: string;
+  lastTimeCheck: number;
+  [key: string]: string | number;
+}
+
+export function showNotification(newVersion: string): void {
+  // TODO: Move inside UpdateServiceBanner
+  alert(newVersion);
+}
+
+function getVersionFromVersionObject(versionObject: any): string {
+  return Object.keys(versionObject.response.results)[0];
```

Comment:
I think we have TS misconfigured, this should scream so much at you for accessing unknowns. What should this return if the versionsObject does not have a response or the response has no results or the results is an empty array? Some of those will return no string but a null, others will throw an error, which means that the type said to be returned here is wrong.

Reply:
I added `interface VersionObject`, but then I got
![screenshot from 2018-11-20 14-47-18](https://user-images.githubusercontent.com/40791275/48774585-539f3f00-ecd3-11e8-9ddc-7d2f911fa54c.png)
.
And when I changed `RequestResponse`,
![screenshot from 2018-11-20 14-50-05](https://user-images.githubusercontent.com/40791275/48774720-aa0c7d80-ecd3-11e8-8a37-20e5ca740ea6.png)
. Observables with types are weird.

Reply:
So let me summarize what typescript tells you:

You told typescript that the request always returns an object on which the response key is exactly an empty object (by using `Observable<RequestResponse<{}>>`) and now you try to access on this empty object a result key, which is not there for the empty object, therefore the error.

If you set the `Observable<RequestResponse<{}>>` to something closer to the truth, maybe `Observable<RequestResponse<CosmosListResponse>>` and you give the `CosmosListResponse` a key attribute that is structured the same way as the HTTP response is then you should be better of. Keep in mind that if you don't know if a key is really always present you should add a ? `key?: number` or a null as union type `key: number | null`, to ensure that the type assertions are always true. After that, you need to alter the code in the map to be clearer with which types are always there and which ones are not.

If you find it confusing to work with typescript in observables I can recommend to use full functions with type signatures for every map, that way you can be sure of which type the outcome is.